### ISSUE TITLE
on Windows, tail reads lines repeatedly from symlinked source

### ISSUE THREAD

**Issue by krancour** [fluent/fluent-bit#1806]
## Bug Report

**Describe the bug**

On Windows, lines read from a symlink by the tail plugin get read over and over again, ad infinitum. I do not see the same behavior on Linux.

**To Reproduce**

C:\fluent-bit\data\test-data.txt:
```
aaa
bbb
ccc
```

Make a symlink:

```powershell
$ New-Item -Path C:\fluent-bit\data\test-data-link.txt -ItemType SymbolicLink -Value C:\fluent-bit\data\test-data.txt
```

Reference that in some simple configuration:

C:\fluent-bit\conf\fluent-bit.conf:
```
[SERVICE]
    Flush         1
    Log_Level     debug
    Daemon        off

[INPUT]
    Name              tail
    Tag               test
    Path              C:\fluent-bit\data\test-data-link.txt
    Mem_Buf_Limit     5MB
    Skip_Long_Lines   On
    Refresh_Interval  10

[OUTPUT]
    Name          stdout
    Match         test
```

Let 'er rip:

```powershell
$ C:\fluent-bit\bin\fluent-bit.exe -c C:\fluent-bit\conf\fluent-bit.conf
```

**Let this run for a few minutes.**

Observe the following output:
```
Fluent Bit v1.3.3
Copyright (C) Treasure Data

[2019/12/12 05:59:08] [debug] [storage] [cio stream] new stream registered: tail.0
[2019/12/12 05:59:08] [ info] [storage] initializing...
[2019/12/12 05:59:08] [ info] [storage] in-memory
[2019/12/12 05:59:08] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
[2019/12/12 05:59:08] [ info] [engine] started (pid=2768)
[2019/12/12 05:59:08] [debug] [engine] coroutine stack size: 98302 bytes (96.0K)
[2019/12/12 05:59:08] [debug] [in_tail] scanning path c:\fluent-bit\data\test-data-link.txt
[2019/12/12 05:59:08] [debug] [in_tail] add to scan queue c:\fluent-bit\data\test-data-link.txt, offset=0
[2019/12/12 05:59:08] [debug] [router] match rule tail.0:stdout.0
[2019/12/12 05:59:08] [ info] [sp] stream processor started
[2019/12/12 05:59:08] [debug] [in_tail] file=c:\fluent-bit\data\test-data-link.txt read=15 lines=3
[2019/12/12 05:59:08] [debug] [in_tail] file=c:\fluent-bit\data\test-data-link.txt promote to TAIL_EVENT
[2019/12/12 05:59:09] [debug] [task] created task=00000262409EF0F0 id=0 OK
[0] test: [1576130348.092644900, {"log"=>"aaa"}]
[1] test: [1576130348.092663400, {"log"=>"bbb"}]
[2] test: [1576130348.092663800, {"log"=>"ccc"}]
[2019/12/12 05:59:09] [debug] [task] destroy task=00000262409EF0F0 (task_id=0)
[2019/12/12 05:59:18] [debug] [in_tail] append new file: c:\fluent-bit\data\test-data-link.txt
[2019/12/12 05:59:18] [debug] [in_tail] add to scan queue c:\fluent-bit\data\test-data-link.txt, offset=0
[2019/12/12 05:59:18] [debug] [in_tail] file=c:\fluent-bit\data\test-data-link.txt read=15 lines=3
[2019/12/12 05:59:18] [debug] [in_tail] file=c:\fluent-bit\data\test-data-link.txt promote to TAIL_EVENT
[2019/12/12 05:59:19] [debug] [task] created task=00000262409EF0F0 id=0 OK
[0] test: [1576130358.114188100, {"log"=>"aaa"}]
[1] test: [1576130358.114193700, {"log"=>"bbb"}]
[2] test: [1576130358.114194000, {"log"=>"ccc"}]
[2019/12/12 05:59:19] [debug] [task] destroy task=00000262409EF0F0 (task_id=0)
[2019/12/12 05:59:28] [debug] [in_tail] append new file: c:\fluent-bit\data\test-data-link.txt
[2019/12/12 05:59:28] [debug] [in_tail] add to scan queue c:\fluent-bit\data\test-data-link.txt, offset=0
[2019/12/12 05:59:28] [debug] [in_tail] file=c:\fluent-bit\data\test-data-link.txt read=15 lines=3
[2019/12/12 05:59:28] [debug] [in_tail] file=c:\fluent-bit\data\test-data-link.txt promote to TAIL_EVENT
[2019/12/12 05:59:29] [debug] [task] created task=00000262409EF0F0 id=0 OK
[0] test: [1576130368.094498000, {"log"=>"aaa"}]
[1] test: [1576130368.094501900, {"log"=>"bbb"}]
[2] test: [1576130368.094502100, {"log"=>"ccc"}]
[2019/12/12 05:59:29] [debug] [task] destroy task=00000262409EF0F0 (task_id=0)
```

**Expected behavior**

I expect that each line is read only once.

**Your Environment**

* Version used: 1.3.3 

* I've been able to replicate this problem in the following environments:

    * Windows nodes in Kubernetes v1.14.8 on AKS
    * Plain Docker, in containers based on `mcr.microsoft.com/windows/servercore:ltsc2019`
    * A Windows 10 Pro VM

**Additional context**

I'm trying to get fuent-bit cranking on the Windows nodes in my mixed node Kubernetes cluster. Container logs, are of course, symlinks. 🤷‍♂

**Comment by fujimotos** [comment]
> On Windows, lines read from a symlink by the tail plugin get read over and over again, ad infinitum. I do not see the same behavior on Linux.

@krancour I can confirm this is broken in HEAD. Symbolic links are
added to queue on each scan as if they are new files.

```
C:\fluent-bit> mklink link.txt log.txt
C:\fluent-bit> fluent-bit.exe -i tail -p path=link.txt -p refresh_interval=1 -o stdout -v
...
[2019/12/12 16:25:27] [debug] [in_tail] scanning path link.txt
[2019/12/12 16:25:27] [debug] [in_tail] add to scan queue C:\fluent-bit\link.txt, offset=0
[2019/12/12 16:25:27] [trace] [in_tail] path: link.txt
[2019/12/12 16:25:27] [debug] [router] default match rule tail.0:stdout.0
[2019/12/12 16:25:27] [debug] [in_tail] file=C:\fluent-bit\link.txt read=8 lines=1
[2019/12/12 16:25:27] [debug] [in_tail] file=C:\fluent-bit\link.txt promote to TAIL_EVENT
[2019/12/12 16:25:28] [debug] [in_tail] append new file: C:\fluent-bit\link.txt
[2019/12/12 16:25:28] [debug] [in_tail] add to scan queue C:\fluent-bit\link.txt, offset=0
[2019/12/12 16:25:28] [debug] [in_tail] file=C:\fluent-bit\link.txt read=8 lines=1
[2019/12/12 16:25:28] [debug] [in_tail] file=C:\fluent-bit\link.txt promote to TAIL_EVENT
```

Evidently the path normalization logic in `tail_scan_win32.c` is not enough.
Fluent Bit should never add a file to scan queue when it is already tracked.

I'll work on a fix shortly.

**Comment by edsiper** [comment]
cc: @fujimotos

**Comment by krancour** [comment]
@fujimotos I am curious... if the issue is one of improper normalization, wouldn't the paths we see added to the scanner over and over again appear to be different?

**Comment by krancour** [comment]
@fujimotos it's this bit in `tail_file.h` that is to blame:

```
static inline int flb_tail_file_name_cmp(char *name,
                                        struct flb_tail_file *file)
{
#if defined(__linux__)
    return strcmp(name, file->name);
#elif defined(FLB_SYSTEM_WINDOWS)
    return _stricmp(name, file->real_name);
#else
    return strcmp(name, file->real_name);
#endif
}

This is called by `flb_tail_file_append(...)` in determining whether a file needs to be added to the scanner. On Windows, the path (link) is being compared to the `real_name` (link target) and obviously, those never match, and `flb_tail_file_append(...)` decides _every_ time it is called on a given symlnk that it must be added.

So there's the problem.

In terms of straightening it out, I defer to you. I'm not a strong C guy and on top of that, I've noticed some ripples effects if I just change `flb_tail_file_name_cmp()` to always look at `name` instead of `real_name`. I don't think the fix is as simple as that.

Anyway... just wanted to share what I'd figured out in case it helps lead to faster resolution.

Thanks again for looking into this.
```

**Comment by krancour** [comment]
Ok... I think I have it all worked out now. PR to follow shortly.
